gradle-plugin/src/main/groovy/org/groovymc/modsdotgroovy/gradle/tasks/GatherNeoForgePlatformDetails.groovy

```diff
@@ -2,67 +2,44 @@ package org.groovymc.modsdotgroovy.gradle.tasks
-import net.neoforged.gradle.dsl.common.runs.run.Run
-import net.neoforged.gradle.dsl.common.runtime.spec.Specification
-import org.gradle.api.NamedDomainObjectContainer
+import org.gradle.api.provider.Provider
 import org.gradle.api.tasks.CacheableTask
-import org.gradle.api.tasks.TaskAction
 import org.jetbrains.annotations.Nullable
 
+import javax.inject.Inject
+
 @CacheableTask
 @CompileStatic
 abstract class GatherNeoForgePlatformDetails extends AbstractGatherPlatformDetailsTask {
-
-    @TaskAction
-    void run() throws IllegalStateException {
-        @Nullable String minecraftVersion = this.minecraftVersion.getOrNull() ?: versions[0]
-        @Nullable String platformVersion = this.platformVersion.getOrNull() ?: versions[1]
-
-        if (minecraftVersion === null || platformVersion === null)
-            throw new IllegalStateException("""
-                Could not find Minecraft or NeoForge version in run configs for project \"${project.name}\".
-                Try manually setting the minecraftVersion and platformVersion properties for this task.
-            """.strip().stripIndent())
-
-        this.writePlatformDetails(minecraftVersion, platformVersion)
-    }
-
-    @Memoized
-    private @Nullable String[] getVersions() {
-        @Nullable String mcVersion = null
-        @Nullable String neoForgeVersion = null
-
-        final NamedDomainObjectContainer<Run> runs = (NamedDomainObjectContainer<Run>) project.extensions.findByName('runs')
-        for (run in runs) {
-            if (run !instanceof RunImpl) continue
-
-            for (sourceSet in run.modSources.get()) {
-                final @Nullable runtimeDef = (CommonRuntimeDefinition) TaskDependencyUtils.findRuntimeDefinition(project, sourceSet).orElse(null)
-                if (runtimeDef === null) continue
-
-                final Specification spec = runtimeDef.specification
-                mcVersion ?= spec.minecraftVersion
-                neoForgeVersion ?= getForgeVersionFromSpec(spec)
-            }
-        }
-
-        return new String[] { mcVersion, neoForgeVersion }
+    @Inject
+    GatherNeoForgePlatformDetails(String configurationName) {
+        var versions = calculateVersions(configurationName)
+        this.minecraftVersion.convention(versions.map { it[0] })
+        this.platformVersion.convention(versions.map { it[1] })
     }
 
     /**
-     * Due to a NeoGradle 7 bug/limitation, we can't depend on the userdev module to do this statically as it'll cause
-     * crashes for users when the versions don't precisely match up.
+     * Due to gradle classloader limitations, this has to be done dynamically
      */
     @CompileDynamic
-    private static @Nullable String getForgeVersionFromSpec(Specification spec) {
-        try {
-            return spec.getForgeVersion()
-        } catch (e) {
-            return null
+    private @Nullable Provider<String[]> calculateVersions(String configurationName) {
```

Comment:
```suggestion
    private Provider<@Nullable String[]> calculateVersions(String configurationName) {
```
Minor: the Provider itself will never be null - only what it returns.

Reply:
Minor again: It's the array members that are nullable, not the array, so it should be `Provider<String @Nullable []>`

Reply:
(Fixed it though)